src/rpc/services/SpectateService.ts

```diff
@@ -0,0 +1,531 @@
+                if (
+                    snapshotPayload.disputeWindows &&
+                    snapshotPayload.disputeWindows.length > 0
+                ) {
+                    // Pop the next dispute window
+                    const currentWindowDisputes =
+                        snapshotPayload.disputeWindows.shift()!;
+
+                    const creationTimestamp =
+                        await stateManager.stateChannelManagerContract.getDisputeWindowCreationTimestamp(
+                            channelId,
+                            currentForkId
+                        );
+
+                    // Reduce only this window
+                    const reducedOutput =
+                        await stateManager.stateChannelManagerContract.reduceProxyView(
+                            currentWindowDisputes,
+                            creationTimestamp
+                        );
+                    console.log(
+                        `Computed reduced output: forkGenesisTimestamp=${reducedOutput.forkGenesisTimestamp}`
+                    );
+
+                    // Move to next fork
+                    const nextForkId =
+                        reducedOutput.latestBlock.stateSnapshotHash;
+                    currentForkId = nextForkId;
+                    isDisputed =
+                        await stateManager.stateChannelManagerContract.isForkDisputed(
+                            channelId,
+                            currentForkId
+                        );
+                } else {
+                    // No disputes provided, can't proceed
+                    return {
+                        isValid: false,
+                        error: "No disputes provided for disputed fork"
+                    };
+                }
+            }
+
+            // Get the fork genesis snapshot from the participant's payload
+            const participantProvidedSnapshot =
+                snapshotPayload.latestForkGenesisSnapshot;
+
+            // Verify that the computed fork ID matches the participant's claimed fork ID
+            if (currentForkId !== participantProvidedSnapshot.forkId) {
+                return {
+                    isValid: false,
+                    error: `Computed fork ID ${currentForkId} does not match participant provided fork ID ${participantProvidedSnapshot.forkId}`
+                };
+            }
+
+            console.log(`Verified fork ID matches: ${currentForkId}`);
+
+            // Verify state proof to prove the latest state from the proven fork
+            if (
+                snapshotPayload.stateProof &&
+                snapshotPayload.stateProof.milestones.length > 0
+            ) {
+                console.log(`Verifying state proof from proven fork`);
+
+                // Get the latest state proof from the proven fork to compare with participant's state proof
+                const latestBlockHeight =
+                    stateManager.storage.blocks.getNextBlockHeight(
```

Comment:
the verifyier/spectator doesn't have any blocks in storage - to verify the stateProof specificaly the milestones which we're interested in, only the you would reuse the solidity code already defined - see `updateStateSnapshotSameFork` in `StateSnapshotFacet` - the only difference is that `updateStateSnapshotSameFork` reads the `currentStateSnapshot` from storage and you'd pas in `genesisStateSnapshot`, so you'd slighly have to refactor the code to have a function that returns true/false and takes in an additional snapshot as an argument and use that one (that you can pass the genesisSnapshot) here and `updateStateSnapshotSameFork` uses that internally and passes the snapshot read from storage